res/controllers/Traktor-Kontrol-S4-MK3.js

```diff
@@ -0,0 +1,1949 @@
+            }
+            if (this.longPressTimer !== 0) {
+                engine.stopTimer(this.longPressTimer);
+            }
+            this.longPressTimer = 0;
+            this.isLongPressed = false;
+        }
+    }
+}
+
+class PlayButton extends ToggleButton {
+    constructor(options) {
+        super(options);
+        this.inKey = "play";
+        this.outKey = "play_indicator";
+        this.outConnect();
+    }
+}
+
+class CueButton extends PushButton {
+    constructor(options) {
+        super(options);
+        this.outKey = "cue_indicator";
+        this.outConnect();
+    }
+    unshift() {
+        this.inKey = "cue_default";
+    }
+    shift() {
+        this.inKey = "start_stop";
+    }
+}
+
+class Encoder extends Component {
+    constructor(options) {
+        super(options);
+        this.lastValue = null;
+    }
+    isRightTurn(value) {
+        // detect wrap around
+        const oldValue = this.lastValue;
+        this.lastValue = value;
+        if (oldValue === this.max && value === 0) {
+            return true;
+        }
+        if (oldValue === 0 && value === this.max) {
+            return false;
+        }
+        return value > oldValue;
+    }
+}
+
+class HotcueButton extends PushButton {
+    constructor(options) {
+        super(options);
+        if (this.number === undefined || !Number.isInteger(this.number) || this.number < 1 || this.number > 32) {
+            throw Error("HotcueButton must have a number property of an integer between 1 and 32");
+        }
+        this.outKey = "hotcue_" + this.number + "_enabled";
+        this.colorKey = "hotcue_" + this.number + "_color";
+        this.outConnect();
+    }
+    unshift() {
+        this.inKey = "hotcue_" + this.number + "_activate";
+    }
+    shift() {
+        this.inKey = "hotcue_" + this.number + "_clear";
+    }
+    output(value) {
+        if (value) {
+            this.send(this.color + this.brightnessOn);
+        } else {
+            this.send(0);
+        }
+    }
+    outConnect() {
+        if (undefined !== this.group) {
+            this.outConnections[0] = engine.makeConnection(this.group, this.outKey, this.output.bind(this));
+            this.outConnections[1] = engine.makeConnection(this.group, this.colorKey, (colorCode) => {
+                this.color = this.colorMap.getValueForNearestColor(colorCode);
+                this.output(engine.getValue(this.group, this.outKey));
+            });
+        }
+    }
+}
+
+class SamplerButton extends Button {
+    constructor(options) {
+        super(options);
+        if (this.number === undefined || !Number.isInteger(this.number) || this.number < 1 || this.number > 64) {
+            throw Error("SamplerButton must have a number property of an integer between 1 and 64");
+        }
+        this.group = "[Sampler" + this.number + "]";
+        this.outConnect();
+    }
+    input(pressed) {
+        if (!this.shifted) {
```

Comment:
Yes, that is intentional. The old way with MIDI Components of reassigning the `input` method IMO ended up being rather awkward and clumsy in many cases. I do not recall who it was, but I do remember a mapping developer giving feedback on the MIDI Components library and mentioning that approach being overcomplicated. With this system it would be even worse because the input callback would need to be deregistered from the HIDInputPacket and the new one would need to be registered every time shift is pressed.

Reply:
Mhmm. I'll think about that. I find the overwriting input method a bit more elegant, but it also forces the JIT to deoptimize code so it might not be that good idea after all...

Reply:
My apologies for jumping in here, I just read your discussion and would like to add my 2 cents.

Both of you already had a look into the custom [library](https://github.com/mixxxdj/mixxx/blob/2.3/res/controllers/Behringer-Extension-scripts.js) that I contributed with Behringer devices in mind. I chose a different approach for the shift logic there and want to share my thoughts; maybe some details may be applied to other mappings, too.

I decided against the idea of handling the different behaviors on shifted/unshifted state inside a JS Component. Instead, the two behaviors are split into two components which are registered/unregistered according to the state of the shift button, so they are unaware of shifting.

In my implementation, you'll find a shiftable `Component` called "layer manager" which holds 2 component containers (= layers) named `Default` and `Shift`. JS Components may be added to either layer. This layer manager is the only component that knows about shifting. When it is shifted or unshifted, all affected components on the corresponding layer are registered / unregistered. Components that are not shiftable stay untouched.

The benefit of this approach is a good separation of concerns for the cost of more frequent component registrations. In my mappings, the number of shiftable components is small (measured against the total component count), and thus the number of registration changes is low. This approach works fine for me: the components look uncluttered and I didn't notice any effect on the performance.

Reply:
While that is another valid approach, that's not what I want to do. That would require considerably more boilerplate and complexity. Most Components have some alternative functionality with shift so that complexity would add up.

Reply:
I consider the layered approach as the more versatile and cleaner architecture. But it may depend on the use case. Probably both in parallel are needed.

Reply:
The layered approach also better resembles how *Shift* is actually implemented, i.e. it alters the feedback of all controls and not only a single one.

Reply:
I think it does depend on the use case and it is good to support both approaches.

Reply:
The layered approach described seems valid. The actual physical control surface is still a single physical component, so using multiple components seems a bit wrong. We should think about adjusting terminology or maybe introduce a new concept for this. 
Using another container class seems like an interesting concept actually. If that container class is responsible about connecting components based on whether they're shifted or not, it could actually have customizable behavior depending on the controller. For example for taking advantage of the controllers firmware shift feature.